drivers/stepper/adi_tmc/adi_tmc5041_stepper_controller.c

```diff
@@ -0,0 +1,764 @@
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2024 Carl Zeiss Meditec AG
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT adi_tmc5041
+
+#include "stdlib.h"
+
+#include <zephyr/drivers/stepper.h>
+#include <zephyr/drivers/stepper/stepper_trinamic.h>
+
+#include "adi_tmc_reg.h"
+#include "adi_tmc_spi.h"
+
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_REGISTER(tmc5041, CONFIG_STEPPER_LOG_LEVEL);
+
+struct tmc5041_data {
+	struct k_sem sem;
+};
+
+struct tmc5041_config {
+	const uint32_t gconf;
+	struct spi_dt_spec spi;
+	const uint32_t clock_frequency;
+};
+
+struct tmc5041_stepper_data {
+	struct k_work_delayable stallguard_dwork;
+	/* Work item to run the callback in a thread context. */
+#ifdef CONFIG_STEPPER_ADI_TMC5041_RAMPSTAT_POLL
+	struct k_work_delayable rampstat_callback_dwork;
+#endif
+	/* device pointer required to access config in k_work */
+	const struct device *stepper;
+	struct k_poll_signal *async_signal;
+};
+
+struct tmc5041_stepper_config {
+	const uint8_t index;
+	const uint16_t default_micro_step_res;
+	const int8_t sg_threshold;
+	const bool is_sg_enabled;
+	const uint32_t sg_velocity_check_interval_ms;
+	const uint32_t sg_threshold_velocity;
+	/* parent controller required for bus communication */
+	const struct device *controller;
+#ifdef CONFIG_STEPPER_ADI_TMC_RAMP_GEN
+	const struct tmc_ramp_generator_data default_ramp_config;
+#endif
+};
+
+static int tmc5041_write(const struct device *dev, const uint8_t reg_addr, const uint32_t reg_val)
+{
+	const struct tmc5041_config *config = dev->config;
+	struct tmc5041_data *data = dev->data;
+	const struct spi_dt_spec bus = config->spi;
+	int err;
+
+	k_sem_take(&data->sem, K_FOREVER);
+
+	err = tmc_spi_write_register(&bus, TMC5041_WRITE_BIT, reg_addr, reg_val);
+	if (err) {
+		LOG_ERR("Failed to write register 0x%x with value 0x%x", reg_addr, reg_val);
+	}
+	k_sem_give(&data->sem);
+
+	return err;
```

Comment:
`return 0` (wherever you can in the file, that way it's unambiguous that the function termination is always the success case and the errors are branched out)